File: free/NumJewelsInStones.js
/**
 * Given string J distinct, find how many jewels are in string S stones, case sensitive.
 * You would need to effectively iterate through the stones and check if they are a jewel.
 * Create a hash map of the jewels and iterate through each stone and increment a counter.
 * 
 * @param {string} J
 * @param {string} S
 * @return {number}
 */
const numJewelsInStones = (J, S) => {
	let jewelCount = 0;

	let jewelMap = new Map();
	for (let jewel of J) {
		console.log(jewel);
		jewelMap.set(jewel, true);
	}

	for (let stone of S) {
		jewelMap.get(stone) && jewelCount++;
	}

	return jewelCount;
};

File: explore/hash_table/num_jewels_in_stones.py
# Given strings j jewels, and s stones, find how many stones are also jewels. All jewels are distinct and everything is case sensitive.
# Pretty simple. Make a boolean hash map for each jewel then iterate through stones and count sum for each jewel encounter.
class Solution:
	def numJewelsInStones(self, J: str, S: str) -> int:
		# edge case
		if not J or not S:
			return 0

		# populate jewel hash
		jewel_hash = {}
		for jewel in J:
			if jewel not in jewel_hash:
				jewel_hash[jewel] = True

		# iterate through stones and build jewel count
		jewel_count = 0
		for stone in S:
			if stone in jewel_hash:
				jewel_count += 1

		return jewel_count